graspologic/partition/leiden.py

```diff
@@ -0,0 +1,402 @@
+            "to be directed"
+        )
+
+    return utils.to_weighted_edge_list(
+        graph=graph,
+        weight_attribute=weight_attribute,
+        weight_default=weight_default,
+        is_weighted=is_weighted,
+        is_directed=False,
+    )
+
+
+def _validate_common_arguments(
+    starting_communities: Optional[Dict[str, int]] = None,
+    iterations: int = 1,
+    resolution: float = 1.0,
+    randomness: float = 0.001,
+    use_modularity: bool = True,
+    random_seed: Optional[int] = None,
+    is_weighted: Optional[bool] = None,
+    weight_default: float = 1.0,
+    check_directed: bool = True,
+):
+    if starting_communities is not None and not isinstance(starting_communities, dict):
+        raise TypeError("starting_communities must be a dictionary")
+    if not isinstance(iterations, int):
+        raise TypeError("iterations must be an int")
+    if not isinstance(resolution, (int, float)):
+        raise TypeError("resolution must be a float")
+    if not isinstance(randomness, (int, float)):
+        raise TypeError("randomness must be a float")
+    if not isinstance(use_modularity, bool):
+        raise TypeError("use_modularity must be a bool")
+    if random_seed is not None and not isinstance(random_seed, int):
+        raise TypeError("random_seed must either be an int or None")
+    if is_weighted is not None and not isinstance(is_weighted, bool):
+        raise TypeError("is_weighted must either be a bool or None")
+    if not isinstance(weight_default, (int, float)):
+        raise TypeError("weight_default must be a float")
+    if not isinstance(check_directed, bool):
+        raise TypeError("check_directed must be a bool")
+
+    if iterations < 1:
+        raise ValueError("iterations must be a positive integer")
+    if resolution <= 0:
+        raise ValueError("resolution must be a positive float")
+    if randomness <= 0:
+        raise ValueError("randomness must be a positive float")
+    if random_seed is not None and random_seed <= 0:
+        raise ValueError(
+            "random_seed must be a positive integer (the native PRNG implementation is an unsigned 64 bit integer)"
+        )
+
+
+def leiden(
+    graph: Union[
+        List[Tuple[Any, Any, Union[int, float]]],
+        nx.Graph,
+        np.ndarray,
+        scipy.sparse.csr.csr_matrix,
+    ],
+    starting_communities: Optional[Dict[str, int]] = None,
+    iterations: int = 1,
+    resolution: float = 1.0,
+    randomness: float = 0.001,
+    use_modularity: bool = True,
+    random_seed: Optional[int] = None,
+    weight_attribute: str = "weight",
+    is_weighted: Optional[bool] = None,
+    weight_default: float = 1.0,
+    check_directed: bool = True,
+) -> Dict[str, int]:
+    """
+    Leiden is a global network partitioning algorithm. Given a list of edges and a maximization function, it will
+    iterate through the network attempting to find an optimal partitioning of the entire network.
```

Comment:
This description doesn't really say what it is trying to optimize for - communities that have more connections within community than without i.e. assortative communities. 

Maybe remove "list of edges" and replace with "graph". 

"attempting to find" makes this less of an issue but I almost always try to avoid saying "optimal" unless I mean it in some rigorous sense, here we don't make any guarantees about whether or not this algo will work and under what model.

Reply:
Good catch on list of edges - that's the docstring from Rust which will outright fail if you pass it anything other than a Vec<(usize, usize, f64)> (i.e. a list of edges), and it didn't get updated when we wrote the wrapper.

"attempting to find an optimal model" sounds about right, doesn't it?  We aren't guaranteeing we will, and it's an open question as to what optimal means.  Significantly better than a purely random partitioning, but no guarantees it will be mathematically the best partition structure possible... what phrase would you use instead?